examples/server/server.cpp

```diff
@@ -3163,10 +2593,11 @@ int main(int argc, char **argv)
                 json data = oaicompat_completion_params_parse(json::parse(req.body));
 
                 const int task_id = llama.request_completion(data, false, false, -1);
+                llama.queue_results.add_waiting_task_id(task_id);
 
```

Comment:
I wonder, do we need to explicitly add waiting task ids like this - can it become automatic within `request_completion` or more specifically inside the `post()` method?

Main thing is that I'm worried about this pattern:

```c++
    void add_waiting_task_id(int task_id) {
        std::unique_lock<std::mutex> lock(mutex_results);
        waiting_task_ids.insert(task_id);
    }
```

This makes the `waiting_task_ids` container atomic, but it hides a potential race condition where we haven't added the task id yet, but the receiver is already waiting for it.

Similar thoughts about `remove_waiting_task_id` - I think it would be better to absorb it within `recv()`. What do you think?

Reply:
During my testing, I ran into issue where user cancel the task (disconnect the HTTP stream) while the tasks are still being processed. The results are still being sent even when the socket is already closed (because the "cancel" task is not yet to be processed). The main idea of that `waiting_task_ids` is to keep track if the HTTP stream is still alive or not.

I did try moving both `add_waiting_task_id` and `remove_waiting_task_id` into `recv()`, but the problem is that in streaming mode, what if the result arrive while the HTTP server is busy sending data and that the next `recv()` is not yet called? In that situation, I saw some results to be dropped.

About the race condition that you pointed out, I don't think that will happen. Basically:
- We always call `add_waiting_task_id` before `recv`. This is always sequential
- `remove_waiting_task_id` only get called right before the HTTP stream is closed, so no result is dropped as long as HTTP stream is alive
- My idea is inspired by the idea of socket `accept`, then a `recv` loop and finally `close`

Also, it worth to mention that one HTTP request is linked to one or more tasks, but one task is never be shared among HTTP requests. In the worst case, a result could not be picked up by any of the HTTP requests, it would remand in the result queue all the time. Therefore, I think it's still worth somehow keeping track of the state of HTTP connection in order not to accept "abandoned result".

But it's true that I may messed up somewhere so I saw some strange behavior. I think I'll revisit this part in another PR. What do you think about that @ggerganov ?

Reply:
I could be misreading something, but I think there is definitely a race:

```
thread 0   -> llama.request_completion()
thread   1 -> llama_server_response.send()       <-- does nothing because no waiting tasks
thread 0   -> llama.queue_results.add_waiting_task_id()
thread 0   -> llama.queue_results.recv(task_id); <-- result is missed
```

Reply:
Ah yeah you're right, sorry I missed that point. Thanks for having noticed.

I noticed the same potential issue with `split_multiprompt_task`, where I add all subtasks before adding the multitask, will definitely be fix.

Then my idea would be to generate the task ID using `queue_tasks.get_next_id()`, then pass this id into `request_completion` so that we can know the `task_id` before adding it to the queue.

Another idea is to modify to add a param `bool start = true` to `queue_tasks.post()`. When calling `queue_tasks.post(task, false)`, the task not be started right away. Until we call `queue_tasks.start_task(task_id_or_multitask_id)` then the task (or task**s**) will start. But this is quite a patch & also overkill I think.

What do you think about that?

Reply:
Should be resolved in https://github.com/ggerganov/llama.cpp/pull/5065/commits/8f36df8fc9c82eb2b56257217957a09509170e87

The `request_completion` now requires passing an ID for the new task. When a new request arrives, things will be done in the order below:
1. `llama.queue_tasks.get_new_id()`
2. `llama.queue_results.add_waiting_task_id(task_id)`
3. Add the `task_multi` if needed
4. Finally, add all the `task_server`